src/plugins/shell/shell_impl.cpp

```diff
@@ -0,0 +1,145 @@
+#include "shell_impl.h"
+#include "system.h"
+#include "global_include.h"
+#include "px4_custom_mode.h"
+#include <cmath>
+#include <functional>
+#include <array>
+
+namespace mavsdk {
+
+void ShellImpl::init()
+{
+    using namespace std::placeholders; // for `_1`
+    _parent->register_mavlink_message_handler(
+        MAVLINK_MSG_ID_SERIAL_CONTROL,
+        std::bind(&ShellImpl::process_shell_message, this, _1),
+        this);
+}
+
+void ShellImpl::deinit() {}
+
+void ShellImpl::enable() {}
+
+void ShellImpl::disable() {}
+
+ShellImpl::ShellImpl(System& system) : PluginImplBase(system)
+{
+    _parent->register_plugin(this);
+}
+
+ShellImpl::~ShellImpl()
+{
+    _parent->unregister_plugin(this);
+}
+
+void ShellImpl::shell_message_response_async(Shell::result_callback_t callback)
+{
+    _result_subscription = callback;
+}
+
+void ShellImpl::send_shell_message(const Shell::ShellMessage& shell_message)
+{
+    if (!_parent->is_connected()) {
+        if (_result_subscription) {
```

Comment:
I think you need to create locks for all of the member variables to make the plugin thread-safe. You can probably just grab the lock for most of the action, and you probably need to have something to track if we are currently doing a shell roundtrip, or if we can already send the next thing.